### PR TITLE
Invoke-DbaDbDataGenerator - Test for MaskingType Random to support special masking types

### PR DESCRIPTION
**Pull request by andreasjordan**
## Type of Change

 - [x] Bug fix (non-breaking change, fixes # )
 - [ ] New feature (non-breaking change, adds functionality, fixes # )
 - [ ] Breaking change (effects multiple commands or functionality, fixes # )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1`)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/dataplat/appveyor-lab ?
 - [ ] Unit test is included
 - [ ] Documentation
 - [ ] Build system

### Purpose
I wanted to use the special masking types like zipcode, color, etc.

### Approach
Now we test for MaskingType as well and only if this is "Random" we use `Get-DbaRandomizedValue -DataType $columnMaskInfo.ColumnType`, in all other cases we now use `Get-DbaRandomizedValue -RandomizerType $columnMaskInfo.MaskingType`.

### Commands to test
```
$sqlInstance = 'sql01'
$databaseName = 'randomdata'
$tempPath = 'C:\temp'

$db = New-DbaDatabase -SqlInstance $sqlInstance -Name $databaseName
$db.ExecuteNonQuery("CREATE TABLE dbo.randomdata(id int identity primary key, justastring varchar(10), fullname varchar(100), zipcode varchar(10), city varchar(100))")
$config = New-DbaDbDataGeneratorConfig -SqlInstance $sqlInstance -Database $databaseName -Table 'randomdata' -Path $tempPath -Rows 5

Invoke-DbaDbDataGenerator -SqlInstance $sqlInstance -Database $databaseName -FilePath $config -Confirm:$false
$db.Query("SELECT * FROM dbo.randomdata") | Format-Table

Remove-Item -Path $config
$null = Remove-DbaDatabase -SqlInstance $sqlInstance -Database $databaseName -Confirm:$false
```

Old:
image

New:
image

### CODE REVIEW

**Comment by andreasjordan** [comment]
There is an issue because not all the special types are supported. 

To test that, change the table structure to use "bic" as a column name:
```
$db.ExecuteNonQuery("CREATE TABLE dbo.randomdata(id int identity primary key, bic varchar(10), fullname varchar(100), zipcode varchar(10), city varchar(100))")
```

I will provide a temporary fix in the next commit.

**Comment by andreasjordan** [comment]
@sanderstad Do you have time to have a look? My current idea is to add the target MaskingType and SubType to the columntypes.json so that we don't need that code anymore (but get the information from the json):

https://github.com/dataplat/dbatools/blob/ea5ae5b8f0fbc386578cd4f241edc5059efa0468/functions/New-DbaDbDataGeneratorConfig.ps1#L205-L237

**Comment by andreasjordan** [comment]
The file "columntypes.json" is in folder "datamasking" so I have to ensure that the masking commands still work after changing the file...

**Comment by andreasjordan** [comment]
Ok, first step is taken, looks good.

**Comment by andreasjordan** [comment]
If we use the same `Set-Content` no matter if the file already exists, we don't need the if. But we need the ShouldProcess in any case. So I changed that part of the code as well.

**Comment by andreasjordan** [comment]
Running Invoke-DbaDbDataGenerator with different Locale does not change the used locale because Get-DbaRandomizedValue uses $script:faker. So I changed Invoke-DbaDbDataGenerator to use $script:faker as well.

**Comment by andreasjordan** [comment]
I added the support for two part table names here, because I need it in here for testing all changes together - but I can split this change to seperate PR "New-DbaDbDataGeneratorConfig - Support two part table names".

**Comment by andreasjordan** [comment]
@sanderstad : Do you remeber the reason for itentity columns starting at 2 (instead of 1)?:
```
                    if ($columnobject.InPrimaryKey -and $columnobject.DataType.SqlDataType.ToString().ToLowerInvariant() -notmatch 'date') {
                        $min = 2
                    }
```

**Comment by andreasjordan** [comment]
Testing for MaskingType -eq 'Random' is not perfect, as we now still don't have a chance to use "Ramdom"/"Guid". Will have to think about a different solution...

**Comment by potatoqualitee** [comment]
will merge when sander or someone approves 😊

File: functions/New-DbaDbDataGeneratorConfig.ps1
Comment (sanderstad) [inline functions/New-DbaDbDataGeneratorConfig.ps1:139]: I'm not sure if this makes the table collection any faster. The collection is already there, why would you run another command that would do the same thing?

**Comment by sanderstad** [comment]
> @sanderstad : Do you remember the reason for itentity columns starting at 2 (instead of 1)?:
> 
> ```
>                     if ($columnobject.InPrimaryKey -and $columnobject.DataType.SqlDataType.ToString().ToLowerInvariant() -notmatch 'date') {
>                         $min = 2
>                     }
> ```

I have no idea why that value is 2, mostly because I wrote this over a couple of years ago. I guess there was a reason for it otherwise I wouldn't have done it.....

> @sanderstad : Do you remember the reason for itentity columns starting at 2 (instead of 1)?:
> 
> ```
>                     if ($columnobject.InPrimaryKey -and $columnobject.DataType.SqlDataType.ToString().ToLowerInvariant() -notmatch 'date') {
>                         $min = 2
>                     }
> ```

I'm not 100% sure why I set this value to 2. Maybe it was something during my testing that I ran into problems when a primary would have the next value to be set to 1 and this caused issues with generating data.
We could try to make it `1` and check out if that causes any issues

File: functions/Invoke-DbaDbDataGenerator.ps1
Comment (sanderstad) [inline functions/Invoke-DbaDbDataGenerator.ps1:303]: I agree with remplacing the `+=` with just `=` but it's tidier code if the query end with a `;` to make sure it's compliant for future code when that is mandatory.

Comment (andreasjordan) [reply functions/Invoke-DbaDbDataGenerator.ps1:303]: Ok, agreed. Will change that.

**Comment by andreasjordan** [comment]
> > @sanderstad : Do you remember the reason for itentity columns starting at 2 (instead of 1)?:
> > ```
> >                     if ($columnobject.InPrimaryKey -and $columnobject.DataType.SqlDataType.ToString().ToLowerInvariant() -notmatch 'date') {
> >                         $min = 2
> >                     }
> > ```
> 
> I'm not 100% sure why I set this value to 2. Maybe it was something during my testing that I ran into problems when a primary would have the next value to be set to 1 and this caused issues with generating data. We could try to make it `1` and check out if that causes any issues

The purpose of this code is to have only positive values for numeric primary keys (int and bigint) - correct? Because if we don't set `$min`, it is always `$null` - which will be set to the smallest possible value by `Get-DbaRandomizedValue` (-2147483648 and -9223372036854775808).

So in case we use random values, there is not much difference between having min 1 or 2, but I would add a comment to answer future questions.

I used an identity column as primary key and the first value was 2 - but that was not the 2 here in the code - but a problem with getting the correct next identity value for an empty table - which I fixed including a comment.

So whatever we do with this `$min = 2` - I would leave it up to you to decide. I will leave the code as is for now.

File: functions/New-DbaDbDataGeneratorConfig.ps1
Comment (sanderstad) [reply functions/New-DbaDbDataGeneratorConfig.ps1:139]: Good shout. I agree

**Comment by andreasjordan** [comment]
Just found a new issue: `$supportedDataTypes` contains "text" and "ntext", but `Get-DbaRandomizedValue` has no code for that. So a nullable text column will be empty, a non-nullable column will throw an error.

@sanderstad - Do you know why text and ntext are in the supported datatypes?

**Review by sanderstad — APPROVED** [review]
Good changes. Thank you for taking the time

**Comment by andreasjordan** [comment]
I'm still not sure what's the best way do decide if to use `$columnValue = Get-DbaRandomizedValue -DataType $columnobject.ColumnType` or `$columnValue = Get-DbaRandomizedValue -RandomizerType $columnobject.MaskingType -RandomizerSubtype $columnobject.SubType`. 

What should be the default and in what cases should the other one be used?

**Comment by andreasjordan** [comment]
One more thing: `Get-DbaRandomizedValue` uses `$Precision` for 'decimal', 'float', 'money', 'numeric', 'real'. But `Invoke-DbaDbDataGenerator` does not set this parameter so the default of 2 is used. We should use the precision of the datatype, don't we?

**Comment by andreasjordan** [comment]
There might be a risk that it breaks some special use case out there in the wild. And there are more issues I would like to look at. But I think this is a very good start which enables the use of `$columnValue = Get-DbaRandomizedValue -RandomizerType $columnobject.MaskingType -RandomizerSubtype $columnobject.SubType`. So I'm ok with mergeing and will not provide more commits to this.

**Comment by potatoqualitee** [comment]
Thank you both 💯  We'll keep an eye out for any reports of edge-case impacts.

**Pull request merged**
